Extending XML output, adding JSON output.

okofish opened the pull request:
This looks like a fun project! I extended the XML output in what I assume was your original intent, and I added JSON output. I also include limit and offset URL parameters for pagination.

okofish committed:
Extending XML output implementation

okofish committed:
Adding JSON output

okofish committed:
Converting output dates to ISO-8601

okofish committed:
Refactoring XML output to use associative arrays

okofish committed:
Adding approved column to JSON output

okofish committed:
Adding limit and offset URL parameters to XML and JSON output

okofish committed:
Sort by created then by id

ChrisAshworth commented:
Ha, I love that I got a pull request on this before telling anyone about it.

Other comments forthcoming...

ChrisAshworth commented on site/admin/index.php:
```diff
@@ -78,9 +78,9 @@
+            $query_statement = $db->prepare("SELECT * FROM $TABLE_NAME WHERE approved = ? ORDER BY created,id DESC LIMIT 25 OFFSET ?");
         else
-            $query_statement = $db->prepare("SELECT * FROM $TABLE_NAME ORDER BY created DESC LIMIT 25 OFFSET ?");
+            $query_statement = $db->prepare("SELECT * FROM $TABLE_NAME ORDER BY created DESC,id DESC LIMIT 25 OFFSET ?");
```
worth making these match for consistency? (i.e. both of form "created,id DESC" )

okofish replied:
They should both be created DESC,id DESC. I think the DESC only applies to its immediate left neighbor.

ChrisAshworth commented on site/out/json/index.php:
```diff
@@ -0,0 +1,45 @@
+    }
+    else
+    {
+        $sql = "SELECT id,inputsource,content,fromphone,DATE_FORMAT(created, '%Y-%m-%dT%H:%i:%s0Z') as created,approved FROM $TABLE_NAME ORDER BY id " . $order . " LIMIT " . $limit . " OFFSET " . $offset;
```
should this use a prepared statement to prevent injection attacks?

okofish replied:
It's safe because $offset and $limit are always numeric.

ChrisAshworth commented on site/out/xml/index.php:
```diff
@@ -8,21 +32,20 @@
     else
     {
-        $sql = "SELECT content FROM $TABLE_NAME ORDER BY id DESC LIMIT 1";
+        $sql = "SELECT id,inputsource,content,fromphone,DATE_FORMAT(created, '%Y-%m-%dT%H:%i:%s0Z') as created,approved FROM $TABLE_NAME ORDER BY id " . $order . " LIMIT " . $limit . " OFFSET " . $offset;
```
Also here - prepared statement to prevent attacks?

okofish replied:
Again, safe because $offset and $limit are always numeric, and $order is essentially restricted to one of two values. However, a prepared statement could be used for style consistency.

okofish committed:
Correcting SQL statement

okofish commented on site/out/xml/index.php:
```diff
@@ -1,6 +1,30 @@
 <?php
     require "../../CONFIG.php";
-
+    if (is_numeric($_GET["limit"]) && !empty((int) $_GET["limit"])) {
```
Note the is_numeric

okofish committed:
Changed DB columns `approved` and `used` to unsigned tinyint to save memory.

Unsigned tinyint ranges from 0 to 255, so you can still have more than two states if you want.

ChrisAshworth commented:
Regarding is_numeric keeping it safe: ah, fair enough. Even so, I'd prefer using prepared statements for three reasons that go beyond just matching style:

1) It makes the safety clearer, without needing to consider earlier logic.

2) It makes the safety self-contained, and not susceptible to someone making safety-breaking changes in the future. (e.g.: someone downloads this as a starting point to hack on, makes changes to existing parameters or adds new parameters that don't protect the final SQL statement; harder for that to happen if the final SQL is being processed via a prepared statement).

3) It uses safety-related code for its intended purpose; sticks more closely to the intention of the API.

okofish committed:
Changing DB columns `approved` and `used` to signed tinyint

okofish committed:
Adding approved/pending type selection URL param to text output

okofish commented:
RE: prepared statements
OK, makes sense. Any idea how to use fetch_array with prepared statements? The prepare/bind/execute syntax is rather confusing.

okofish commented:
BTW, if you didn't get it working yet, you can configure Twilio/IFTTT with an HTTP-auth'd URL like this:

```
https://user:password@example.com/switchboard/in/twilio/
```

ChrisAshworth commented:
For the other place I did a prepared query I used:

`$query_statement->bind_result($row_id, $row_created, $row_inputsource, $row_fromphone, $row_content, $row_approved, $row_used);`

And then:

`while ($query_statement->fetch())`

However this is just me poking around in PHP docs, I don't use this language much. :-)

ChrisAshworth commented:
Had time this morning to make the necessary adjustments and merge the results. Thanks Jessie!
